browser/brave_wallet/brave_wallet_service_unittest.cc

```diff
@@ -0,0 +1,669 @@
+
+#include "brave/components/brave_wallet/browser/brave_wallet_service.h"
+
+#include "base/test/bind.h"
+#include "brave/components/brave_wallet/browser/erc_token_list_parser.h"
+#include "brave/components/brave_wallet/browser/erc_token_registry.h"
+#include "brave/components/brave_wallet/browser/pref_names.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "chrome/browser/prefs/browser_prefs.h"
+#include "chrome/test/base/testing_browser_process.h"
+#include "chrome/test/base/testing_profile.h"
+#include "components/prefs/pref_service.h"
+#include "components/sync_preferences/testing_pref_service_syncable.h"
+#include "content/public/test/browser_task_environment.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace {
+
+const char token_list_json[] = R"(
+  {
+   "0x06012c8cf97BEaD5deAe237070F9587f8E7A266d": {
+     "name": "Crypto Kitties",
+     "logo": "CryptoKitties-Kitty-13733.svg",
+     "erc20": false,
+     "erc721": true,
+     "symbol": "CK",
+     "decimals": 0
+   },
+   "0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984": {
+     "name": "Uniswap",
+     "logo": "uni.svg",
+     "erc20": true,
+     "symbol": "UNI",
+     "decimals": 18
+   }
+  })";
+
+}  // namespace
+
+namespace brave_wallet {
+
+class BraveWalletServiceUnitTest : public testing::Test {
+ public:
+  BraveWalletServiceUnitTest()
+      : task_environment_(base::test::TaskEnvironment::TimeSource::MOCK_TIME) {}
+  ~BraveWalletServiceUnitTest() override = default;
+
+ protected:
+  void SetUp() override {
+    TestingProfile::Builder builder;
+    auto prefs =
+        std::make_unique<sync_preferences::TestingPrefServiceSyncable>();
+    RegisterUserProfilePrefs(prefs->registry());
+    builder.SetPrefService(std::move(prefs));
+    profile_ = builder.Build();
+    service_.reset(new BraveWalletService(GetPrefs()));
+
+    auto* registry = ERCTokenRegistry::GetInstance();
+    std::vector<mojom::ERCTokenPtr> input_erc_tokens;
+    ASSERT_TRUE(ParseTokenList(token_list_json, &input_erc_tokens));
+    registry->UpdateTokenList(std::move(input_erc_tokens));
+
+    bool callback_called = false;
+    mojom::ERCTokenPtr token1;
+    GetRegistry()->GetTokenByContract(
+        "0x06012c8cf97BEaD5deAe237070F9587f8E7A266d",
+        base::BindLambdaForTesting([&](mojom::ERCTokenPtr token) {
+          token1_ = std::move(token);
+          callback_called = true;
+        }));
+    base::RunLoop().RunUntilIdle();
+    ASSERT_TRUE(callback_called);
+    ASSERT_EQ(token1_->symbol, "CK");
+
+    callback_called = false;
+    mojom::ERCTokenPtr token2;
+    GetRegistry()->GetTokenByContract(
+        "0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984",
+        base::BindLambdaForTesting([&](mojom::ERCTokenPtr token) {
+          token2_ = std::move(token);
+          callback_called = true;
+        }));
+    base::RunLoop().RunUntilIdle();
+    ASSERT_TRUE(callback_called);
+    ASSERT_EQ(token2_->symbol, "UNI");
```

Comment:
Each of this block can be bound into a reusable test lambda function and capture things you need and pass in the parameters to verify different results. In this way we don't have to duplicate the code to verify async result.
See https://github.com/brave/brave-core/pull/10032/commits/3ca0cc9d7833aab38ae9725e01ffeb94b8a23777#diff-684b3b8a18e897e382ea7c678f11653aebd201e60f5b0ebd975c3ccda23f0bddR1362-R1399

Same rule should apply to rest of the test cases

Reply:
If it can be reused across different test cases, we can move it into member function of BraveWalletServiceUnitTest or just anonymous function

Reply:
Reduced in https://github.com/brave/brave-core/pull/9969/commits/4abaaaf80d41d834042e620ddcb5f5b90d0ac8d9.

Reply:
I intentionally keep EXPECT_* expressions outside so it's easier to see which line fails.